stan/math/prim/mat/prob/bernoulli_logit_glm_lpmf.hpp

```diff
@@ -0,0 +1,171 @@
+     * @param n integer parameter
+     * @param x design matrix
+     * @param beta weight vector
+     * @param alpha bias vector (in logits)
+     * @return log probability or log sum of probabilities
+     * @throw std::domain_error if theta is infinite.
+     * @throw std::invalid_argument if container sizes mismatch.
+     */
+
+    template <bool propto,
+              typename T_n,
+              typename T_x,
+              typename T_beta,
+              typename T_alpha>
+    typename return_type<T_x, T_beta, T_alpha>::type
+    bernoulli_logit_glm_lpmf(const T_n& n,
+                             const T_x& x,
+                             const T_beta& beta,
+                             const T_alpha& alpha) {
+      static const std::string function = "bernoulli_logit_glm_lpmf";
+      typedef typename stan::partials_return_type<T_n,
+                                                  T_x,
+                                                  T_beta,
+                                                  T_alpha>::type
+        T_partials_return;
+
+      using stan::is_constant_struct;
+      using std::exp;
+
+      if (!(stan::length(n) &&
+            stan::length(x) &&
+            stan::length(beta) &&
+            stan::length(alpha)))
+        return 0.0;
+
+      T_partials_return logp(0.0);
+
+      check_bounded(function, "n", n, 0, 1);
+      check_not_nan(function, "Data matrix", x);  
+      check_not_nan(function, "Weight vector", beta);
+      check_not_nan(function, "Bias vector", alpha);
+      check_consistent_sizes(function,
+                             "Random variable", n,
+                             "Bias vector", alpha);
+      // check_consistent_sizes(function,
+      //                       "Random variable", n,
+      //                       "Data matrix", x);  // would this work?
+      //                        maybe add one more check to see that x is
+      //                        compatible with beta?
+
+      if (!include_summand<propto, T_x, T_beta, T_alpha>::value)
+        return 0.0;
+
+      scalar_seq_view<T_n> n_vec(n);
+      scalar_seq_view<T_beta> beta_vec(beta);
+      scalar_seq_view<T_alpha> alpha_vec(alpha);
+      
+      const size_t N = max_size(n, alpha);
+      // do we also want to check
+      // anything about x here? and in the next line?
+      const size_t M = beta.size();
+      
+      operands_and_partials<T_x, T_beta, T_alpha> ops_partials(x, beta, alpha);
+      
+      const bool constant_x = is_constant_struct<T_x>::value;
+      const bool constant_beta = is_constant_struct<T_beta>::value;
+      const bool constant_alpha = is_constant_struct<T_alpha>::value;
+      
+      Matrix<double, Dynamic, 1> beta_dbl(M, 1);
+      Matrix<double, Dynamic, Dynamic> x_dbl(N, M);
+      for (size_t m = 0; m < M; m++) {
+        beta_dbl[m] = value_of(beta_vec[m]);
+      }
+
+      for (size_t n = 0; n < N; n++) {  // could we vectorise this loop?
```

Comment:
I took a look at `prim/mat/fun/value_of.hpp` and it seems like you can, in fact, just use `value_of` directly on the matrix and get `x_dbl` back. Technically if this is used in forward mode `x_dbl` could consist of `var`s...

Reply:
Good point on x_dbl possibly consisting of vars if we do higher-order autodiff. I've now changed the types of beta_dbl and x_dbl accordingly. I have opted against calling value_of on x and beta_vec directly as beta_vec is not an Eigen::Matrix, but rather one of these scalar_seq_views. Also, from the implementation of value_of, it seems to me like would pull out all the doubles each time I call the function, even if I only use some of them through indexing.

I feel that I'll need to loop over n anyway to do the Taylor approximation for extreme values in the same way as bernoulli_logit_lpmf, so I might as well keep it as is, without trying to vectorize everything. I appreciate that it'd potentially be slightly nicer if the values of the x matrix did not have to be copied into x_dbl, but it seems like manageable cost. The new function is still much faster than the existing code (by a factor ~7, since I've tuned it a bit).